### PR TITLE
Splitting ModeSolver.solve in some helper functions

### PR DESCRIPTION
**Pull request by momchil-flex**
Also:
- Made most of the helper functions and properties private to not crowd docs
- Added type annotation everywhere

### CODE REVIEW

File: tidy3d/plugins/mode/mode_solver.py [hunk 1]
```diff
@@ -308,35 +308,11 @@ def solve(self) -> ModeSolverData:
...
-                    fields_freq[field_name].append(np.copy(field))
-
-            for field_name, field in fields_freq.items():
-                fields[field_name].append(np.stack(field, axis=-1))
+        n_complex, fields = self._solve_all_freqs(coords=solver_coords, symmetry=solver_symmetry)
 
         # Generate the dictionary of ScalarModeFieldData for every field
         data_dict = {}
-        for field_name, field in fields.items():
+        for field_name in fields[0].keys():
```
Comment (tylerflex) [inline tidy3d/plugins/mode/mode_solver.py:315]: what's the logic behind the [0] here?

**Review by tylerflex — APPROVED** [review]
Just a couple clarifying questions, overall looks good.

File: tidy3d/plugins/mode/mode_solver.py (hunk 1 above)
Comment (momchil-flex) [reply tidy3d/plugins/mode/mode_solver.py:315]: `fields` is a List of Dicts, one for each frequency. I guess we should make sure `self.freqs` is not an empty list?

Comment (tylerflex) [reply tidy3d/plugins/mode/mode_solver.py:315]: yea or maybe check `len(fields)` to make sure it's 1.

**Pull request merged**